chia/wallet/nft_wallet/nft_off_chain.py

```diff
@@ -0,0 +1,96 @@
+import logging
+from pathlib import Path
+from typing import Optional
+
+import requests
+
+from chia.types.blockchain_format.program import Program
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.default_root import DEFAULT_ROOT_PATH
+from chia.util.hash import std_hash
+from chia.wallet.nft_wallet.nft_info import NFTCoinInfo
+from chia.wallet.nft_wallet.uncurry_nft import UncurriedNFT
+
+PREFIX_HASH_LENGTH = 3
+CACHE_PATH_KEY = "nft_metadata_cache_path"
+DEFAULT_PATH = DEFAULT_ROOT_PATH / "nft_cache"
+DEFAULT_PATH.mkdir(parents=True, exist_ok=True)
+log = logging.getLogger(__name__)
+
+
+def fetch_off_chain_metadata(nft_coin_info: NFTCoinInfo) -> Optional[str]:
+    uncurried_nft: Optional[UncurriedNFT] = UncurriedNFT.uncurry(*nft_coin_info.full_puzzle.uncurry())
+    if uncurried_nft is None:
+        return None
+    for uri in uncurried_nft.meta_uris.as_python():  # pylint: disable=E1133
+        response = requests.get(uri)
+        if response.status_code == 200:
+            return response.text
+    return None
+
+
+def read_off_chain_metadata(nft_coin_info: NFTCoinInfo, cache_path: Optional[str] = None) -> Optional[str]:
+    # Read metadata from disk cache
+    cache = DEFAULT_PATH
+    if cache_path is not None:
+        cache = Path(cache_path)
+    file_name = cache / nft_coin_info.nft_id.hex()[:PREFIX_HASH_LENGTH] / nft_coin_info.nft_id.hex()
+    if not file_name.exists():
+        return None
+    try:
+        text = file_name.read_text()
+        if verify_metadata(text, nft_coin_info.full_puzzle):
+            return text
+        else:
+            return None
+    except Exception:
+        return None
```

Comment:
Log the exception? If the file exists and we fail to read it, it might be useful info.

Reply:
What are the possible exceptions here?  We certainly don't want to consume `NameError` or such.

Reply:
Why we don't want to catch NameError? It will reload metadata no matter what kind of exception.

Reply:
I'm happy with the logging addition. I'm not sure I understand what the `NameError` discussion is about @altendky

Reply:
You don't want to catch `NameError` because it indicates a variable name typo or such.  That should fail out hard so we see it quickly and fix it, not hide in the logs where we forget to look while developing and only hear about it when a user sends us a log asking about it.  This is a general commentary on us catching all exceptions in lots of places without thinking about, or often even being aware of, all the relevant exceptions.  `NameError` in particular is just an example.

Reply:
What if we just re-raise the exception here after logging it?

Reply:
have opened a new issue re catching generic exception

Reply:
This in particular was mostly just a discussion.  Reraising is an thing but sometimes we end up double logging then.  Unexpected exceptions should generally bubble really far up and then be reported clearly to developers.  None of the `except Exception:` logging here provides _any_ information about what happened.